src/app/tests/TestPowerSourceCluster.cpp

```diff
@@ -0,0 +1,266 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/support/CHIPMem.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/clusters/power-source-server/power-source-server.h>
+#include <lib/core/TLV.h>
+#include <lib/core/TLVDebug.h>
+#include <lib/core/TLVUtilities.h>
+#include <lib/support/CHIPCounter.h>
+#include <lib/support/ErrorStr.h>
+#include <lib/support/UnitTestContext.h>
+#include <lib/support/UnitTestRegistration.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/Flags.h>
+#include <nlunit-test.h>
+#include <protocols/interaction_model/Constants.h>
+#include <type_traits>
+
+#include <vector>
+
+namespace chip {
+namespace app {
+
+class TestPowerSourceCluster
+{
+public:
+    static void TestEndpointList(nlTestSuite * apSuite, void * apContext);
+};
+
+std::vector<EndpointId> ReadEndpointsThroughAttributeReader(nlTestSuite * apSuite, EndpointId endpoint)
+{
+    Clusters::PowerSourceAttrAccess & attrAccess = Clusters::TestOnlyGetPowerSourceAttrAccess();
+    CHIP_ERROR err                               = CHIP_NO_ERROR;
+
+    // Write TLV through the attribute access interface into the buffer
+    constexpr size_t buflen = 128;
+    uint8_t buf[buflen];
+    memset(buf, 0, buflen);
+    AttributeReportIBs::Builder builder;
+    TLV::TLVWriter tlvWriter;
```

Comment:
Please add empty lines around blocks of related call to help readability of flow.